vortex-alp/src/array.rs

```diff
@@ -43,120 +35,88 @@ impl ALPArray {
+        self.array()
+            .child(0, &self.metadata().encoded_dtype)
+            .expect("Missing encoded array")
     }
 
     pub fn exponents(&self) -> &Exponents {
-        &self.exponents
+        &self.metadata().exponents
     }
 
-    pub fn patches(&self) -> Option<&ArrayRef> {
-        self.patches.as_ref()
+    pub fn patches(&self) -> Option<Array> {
+        self.metadata().has_patches.then(|| {
+            self.array()
+                .child(1, self.dtype())
+                .expect("Missing patches with present metadata flag")
+        })
     }
 }
 
-impl Array for ALPArray {
-    impl_array!();
-
-    #[inline]
-    fn len(&self) -> usize {
-        self.encoded.len()
-    }
-
-    #[inline]
-    fn is_empty(&self) -> bool {
-        self.encoded.is_empty()
-    }
-
-    #[inline]
-    fn dtype(&self) -> &DType {
-        &self.dtype
-    }
-
-    #[inline]
-    fn stats(&self) -> Stats {
-        Stats::new(&self.stats, self)
-    }
-
-    #[inline]
-    fn with_compute_mut(
-        &self,
-        f: &mut dyn FnMut(&dyn ArrayCompute) -> VortexResult<()>,
-    ) -> VortexResult<()> {
-        f(self)
-    }
-
-    #[inline]
-    fn encoding(&self) -> EncodingRef {
-        &ALPEncoding
-    }
-
-    #[inline]
-    fn nbytes(&self) -> usize {
-        self.encoded().nbytes() + self.patches().map(|p| p.nbytes()).unwrap_or(0)
-    }
-
-    fn serde(&self) -> Option<&dyn ArraySerde> {
-        Some(self)
+impl ArrayValidity for ALPArray<'_> {
+    fn is_valid(&self, index: usize) -> bool {
+        self.encoded().with_dyn(|a| a.is_valid(index))
     }
 
-    fn walk(&self, walker: &mut dyn ArrayWalker) -> VortexResult<()> {
-        walker.visit_child(self.encoded())
+    fn logical_validity(&self) -> LogicalValidity {
+        self.encoded().with_dyn(|a| a.logical_validity())
     }
 }
 
-impl ArrayDisplay for ALPArray {
-    fn fmt(&self, f: &mut ArrayFormatter) -> std::fmt::Result {
-        f.property("exponents", format!("{:?}", self.exponents()))?;
-        f.child("encoded", self.encoded())?;
-        f.maybe_child("patches", self.patches())
+impl ArrayFlatten for ALPArray<'_> {
+    fn flatten<'a>(self) -> VortexResult<Flattened<'a>>
+    where
+        Self: 'a,
+    {
+        decompress(self).map(Flattened::Primitive)
     }
 }
 
-impl ArrayValidity for ALPArray {
-    fn logical_validity(&self) -> Validity {
-        self.encoded().logical_validity()
-    }
-
-    fn is_valid(&self, index: usize) -> bool {
-        self.encoded().is_valid(index)
+impl AcceptArrayVisitor for ALPArray<'_> {
+    fn accept(&self, visitor: &mut dyn ArrayVisitor) -> VortexResult<()> {
+        visitor.visit_child("encoded", &self.encoded())?;
+        if self.patches().is_some() {
+            visitor.visit_child(
+                "patches",
+                &self.patches().expect("Expected patches to be present "),
+            )?;
+        }
+        visitor.visit_validity(&self.logical_validity().into_validity())
```

Comment:
do you have to visit this here explicitly if validity is derived from underlying array? I thought this would be unnecessary since this is not stored array